### DISCUSSION TITLE
Client disconnects after sending data

### DISCUSSION THREAD

**Discussion by tobias-carlbom** [jchristn/WatsonTcp#229] [Q&A]
hello again!

I am experiencing a weird problem. When a client sends a message to the server e.g client.Send(“hello”) it just disconnects and returns no errors.

I have try/catches around the send command and also in the server project, messagereceived but no errors. 

At the server project the client donnected and disconnected callbacks are triggered as expected.

**Comment by jchristn** [comment]
Hi @manprinsen can you attach a handler to the client and server ```ExceptionEncountered``` events, and also set ```Settings.DebugMessages``` to true and attach a handler to ```Settings.Logger``` (again on both client and server)?  This should give some insight into what's happening.

**Reply by tobias-carlbom** [reply]
Will do :) brb

**Reply by tobias-carlbom** [reply]
I get the error below at the server project. This gets me thinking if the application that i am trying to integrate this to does not support >4.0.4.1 of the  System.Runtime.CompilerServices.Unsafe.

```2022-11-30 15:30:38 ::: Client connected: 100.0.0.3:53573
MyLogger: Error: [WatsonTcpServer] failed to write message to [c9272e6a-ed3b-4585-bab1-151b2bf5b952|100.0.0.3:53573]: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
MyLogger: Error: Event handler exception in ClientConnected: 
System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
File name: 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.Span`1..ctor(T[] array, Int32 start, Int32 length)
   at System.Text.Json.PooledByteBufferWriter.ClearAndReturnBuffers()
   at System.Text.Json.Utf8JsonWriterCache.ReturnWriterAndBuffer(Utf8JsonWriter writer, PooledByteBufferWriter bufferWriter)
   at System.Text.Json.JsonSerializer.WriteString[TValue](TValue& value, JsonTypeInfo`1 jsonTypeInfo)
   at WatsonTcp.DefaultSerializationHelper.SerializeJson(Object obj, Boolean pretty)
   at WatsonTcp.WatsonTcpServer.SendInternal(ClientMetadata client, WatsonMessage msg, Int64 contentLength, Stream stream)
   at WatsonTcp.WatsonTcpServer.Send(Guid guid, Int64 contentLength, Stream stream, Dictionary`2 metadata)
   at WatsonTcp.WatsonTcpServer.Send(Guid guid, Byte[] data, Dictionary`2 metadata, Int32 start)
   at WatsonTcp.WatsonTcpServer.Send(Guid guid, String data, Dictionary`2 metadata)
   at IoTrack.TcpServer.ClientConnected(Object sender, ConnectionEventArgs args) in C:\Users\Administrator\SynologyDrive\customers\elajo\IoTrack\IoTrack_1.0.5.0_merge_and_tcp_client_server\IoTrack\IoTrack\TcpServer.cs:line 166
   at WatsonTcp.WatsonTcpServerEvents.WrappedEventHandler(Action action, String handler, Object sender)
```

**Comment by jchristn** [comment]
Are you using the NuGet package?  Which operating system are you using?  What runtime (.NET Framework, .NET Core, etc) and which version are you using?  Could you try the steps here?  https://stackoverflow.com/questions/62764744/could-not-load-file-or-assembly-system-runtime-compilerservices-unsafe

**Reply by tobias-carlbom** [reply]
yes i do :) for the watsontcp im running version 5.0.6 and System.Runtime.CompilerServices.Unsafe is version 6.0. if i downgrade the nutget of System.Runtime.CompilerServices.Unsafe to 4.7.1 it seems to work. This also downgrades the watsontcp so i lose a bit of functionallity. As i understand it is not possible to stream, but i might be wrong :)

As for my system: im running on a windows server 2019 and the watsontcp (server) application is created in a class library (.net framwork 4.7.2).
Dont know if .net 4.7.2 is old but this might be the problem

**Reply by jchristn** [reply]
Hi, that's very strange.  I just tested in .NET Framework 4.7.2 and had no issue.  System.Text.Json has the dependency on System.Runtime.CompilerServices.Unsafe (>=6.0.0).  Can you try finding/copying the file in question into the folder manually?  Perhaps it is a packaging issue?

**Reply by tobias-carlbom** [reply]
It is weird :) my conclusion so far is that the tcpserver runs without errors in an console application (latest Watson version) => tcpserver is still in a class library so there’s only a reference in the console app. Client can now send message to server.

But when I integrate the class library to my real application the tcpserver do listen and a client can connect, but the client cannot send data. I’ve tried all versions of version 5 (also changed the version of Watson of the client project).

When I run the latest Watson version of version 4.x it all seems to work, meaning client can send data to server.

I get this message in the log file just before the client disconnects.
MyLogger: Debug: [WatsonTcpServer] starting data receiver for [27180cb5-48be-494d-97df-65002c57ac51|127.0.0.1:54437]
